### ISSUE TITLE
edge call from host to enclave

### ISSUE THREAD

**Issue by asyarifstudio** [keystone-enclave/keystone#329]
Is there any particular reason why the opposite edge call from the host to the enclave, so-called "inbound call", is not implemented? It seems that `enclave.run()` will block the runtime of the host in which it's not able to perform an edge call unless the enclave exits its runtime, in which no more enclave is ready to receive the edge call. 

or there is actually a plan to implement the "inbound call" in the future. what would be the use cases?

**Comment by dayeol** [comment]
I'm assuming that you're referring to something like `ECALL` in SGX. 
Keystone's programming model is a bit different from SGX; the granularity of enclave is a process, not a function, so Keystone should always run an entire process with its own memory space. This is more similar to Gramine (i.e., libOS-based frameworks), where each of the thread runs a single `ECALL` and never returns.
If you want to implement something like gRPC (where an external entity wants to invoke a function inside an enclave), you can technically implement it in application layer; so by design, there's no strong need for having `ECALL` in our programming model.

Implementing gRPC-like service is possible without `ECALL` -- but would need some engineering.

**Comment by asyarifstudio** [comment]
yes. I was thinking something similar with gRPC. I think there could some interesting usecase where host application just need an enclave to perform specific task at any time. Rather recreating the enclave everytime, it can just create an enclave and invoke its RPC method on demand. 

for example, host may create an enclave and load a neural network, and call the enclave through `ECALL` send an input to enclave and wait for the output. other usecase is cryptography where enclave is created to perform cryptographic operation only. Host can also create multiple enclave for different purposes and invoke each gRPC method as needed. 

I was wondering thought if this has been considered by keystone team before? 

anyway thanks for the quick response :)

**Comment by dayeol** [comment]
Yes, it would be really good use case of Keystone.
Would you be interested in implementing it? :D We could talk about what we need to implement and possibly support it with minimal modifications.

**Comment by asyarifstudio** [comment]
I will be looking into it first. I'll you guys know :) but thanks a lot for the support

**Comment by acaldaya** [comment]
there is an alternative Keystone SDK that implements host-to-enclave edge calls using the ocall interface. You can check the details in the repo and paper below:

- https://github.com/vector-sdk/rust-sdk/
- https://cris.vtt.fi/en/publications/towards-a-rust-sdk-for-keystone-enclave-application-development
